wsd: reset the modified flag when saving unmodified doc

Ashod opened the pull request:
In case saving can't really be done (f.e. the doc
is unmodified and there is nothing to save), we
must recognize that we are in fact up-to-date
and there is nothing to do.

Unfortunately this wasn't the case. When we issued
a save request, we only considered it done when it
succeeded. When it failed, we assumed we will
have to try again, until we succeed.

This logic is valid in most normal cases, except
when the document is unmodified (but we think
it *might* be modified).

Here we make sure that we update the last
save attempt so we don't think there might
be some modifications, which we track by
looking at the last activity timestamp compared
to the last save attempt.

This makes it possible to sync-up with Core
when it tells us that there was nothing to save
as the document isn't modified (since the last
save), and so we stop retrying to save an
unmodified document endlessly, which can hold
off an unloading document until timeout, and
even then log a scary warning that modifications
may be lost etc.).

Change-Id: Ia9a91dbc46b2e5b8e08e7caba6472b2315de7afa
Signed-off-by: Ashod Nakashian <ashod.nakashian@collabora.co.uk>

Ashod commented:
Would be best to have #1091 merged before this, as it does improve the stability of the failing test.

Ashod commented:
@gmasei11 would be great if you could take a look at this patch, as you have made changes to the `modified` state tracking in the past.

By the way, will change in a proper state machine soon. So it's unlikely that this change will make it into a release before the state machine, which should be a significant improvement.

gmasei11 reviewed (commented):
I've took a look over this pull request and I have the following concerns:

1. If we put aside your change from this pull request then _isModified flag is synchronized with core unless the upload to storage fails. This distinction is made here: https://github.com/CollaboraOnline/online/blob/0fb1b3376913cc7ae9b2030c81e7a674bef77b27/wsd/ClientSession.cpp#L1346 and it was added in this committ: https://github.com/CollaboraOnline/online/commit/8c602e179e8afe44f2a1fe513bdc7640cd57ed2a. The _isModified flag is modified only here: https://github.com/CollaboraOnline/online/blob/0fb1b3376913cc7ae9b2030c81e7a674bef77b27/wsd/ClientSession.cpp#L1349. Correct me if I'm wrong. Let's suppose that I didn't miss anything. In this case the _isModified flag will remain set until a successful upload occurs. In fact it will remain set until a successful upload followed by a "statechanged" message from core. If only a successful upload is performed then the flag is not reset because the reset is made only on "statechanged" message received from core and the last upload was successful.
2. I'm afraid that the fix that you provide will break "the need of uploading to storage until it is successful". I'm not 100% sure but that's what I'm thinking now, because it seems that it doesn't take into account the case of upload failure.
3. Normally the problem can be solved by resetting the _isModified flag after a successful upload to storage (because we don't reset it if an unsuccessful upload occurs). But this will generate some synchronization problems between _isModified flag and core which will trigger some issues related to autosave fixed by this commit: https://github.com/CollaboraOnline/online/commit/762ca3d3ed613da7a3132192fe535233e1547a9e
4. I was surprised to see that the _isModified flag from DocumentBroker is not synchronized with core. In my opinion we should not break this synchronization. If we want that uploading to storage to affect the modified status from user perspective then we should use a set of flags and combine them only when/where is needed. Something like _isModifiedLocal (replaces current _isModified and reflects core), _isModifiedStorage (only for uploading to storage; should be set only when a need of saving to storage arise - after a local save; should be reset only after a successful upload to storage) and a property named IsModified() that combines the two (_isModifiedLocal || _isModifiedStorage) and these flags should be used whenever the need arise. For example: we should not save locally unless _isModifiedLocal is true or the saving is forced; we should not upload to storage unless _isModifiedStorage is true or the saving is forced; the state from user perspective (UI) should reflect the IsModified() property.

Ashod commented:
Hi @gmasei11, thanks for the detailed feedback. Much appreciated.

It's true that historically the modified flag and the upload state were one and the same. This was a source of problems.

At least since we added the `_lastStorageSaveSuccessful` the upload state and the modified state are separate. So I think your concerns are not an issue.

> Let's suppose that I didn't miss anything. In this case the _isModified flag will remain set until a successful upload occurs.

Not really. `_isModified` will be reset to false when *core* saves the document to disk and sends us "statechanged" with ".uno:ModifiedStatus=true". Uploading is completely irrelevant to this flag and doesn't affect it (Core doesn't care what we do with the saved document, it has no concept of uploading). Notice that my change in this PR sets `_isModified` when *handling the save results from core*. It is *before* the uploading. In fact, it is in the case where we *skip* uploading because the document is not modified (as Core told us so, based on the check in the `if` conditional `result == "unmodified"`).

> I'm afraid that the fix that you provide will break "the need of uploading to storage until it is successful".

I see your point, but my change can only help here! See, Core is telling us the document is *not* modified (i.e. nothing to upload), and my change fixes *two* issues. One, is that we weren't updating the `_isModified` when Core just told us the result of the save. At a minimum I expect we set it to false in this line I added. The second issue is that there is a race between new user input and this same line where we handle the save results from Core. Hence the reason why I check the time of the last activity from the user to see if there is any need to another save (i.e. `if _lastSaveRequestTime < _lastActivityTime` then we assume `_isModified` is true, so we try to save again in the future).

But these issues are only seen when shutting down a document. As I say in the commit message, if we think the document is modified, but Core says it isn't, we end up trying to save and upload until we timeout. That's because we don't update `_isModified` (the first bug I mentioned in the previous paragraph). (Remember that Core will send us "statechanged" with ".uno:ModifiedStatus" only when the modified state changes, which it won't when we are shutting down the document, so `_isModified` will be the last value, and if that was `true`, but Core says nothing to save, then we are stuck. This is the case I fix here.)

Agreed on point 3 and 4: we shouldn't update `_isModified` based on the *upload*, but rather it must be synchronized with Core. This is what my patch does. It makes sure that when Core says the doc isn't modified, we update the `_isModified` flag, but in a smart way (i.e. not setting it to hard-coded false, but seeing if there are any activity from the user as a hint for modifications).

If we think the user made some changes (because there was activity since last save) but in fact there wasn't any change (say, they only moved the mouse), then we will issue another save, which will return "unmodified" and this time we shouldn't have any activities since this second save attempt, and therefore we will set `_isModified` to false (thanks to this PR).

Right now we update `_isModified` in two places: when core sends "statechanged" and (with this PR) when core says the document isn't modified (and fails to save). I'm not worried that my new `setModified` call will conflict with the "statechanged" case, because `_lastSaveRequestTime < _lastActivityTime` is conservative (meaning, it will flag `_isModified` even when it isn't, which will only cost us an extra save attempt, but it should never set the `_isModified` flag to `false` incorrectly. That would be a bad thing, of course.)

Hope this clarifies things and puts your concerns to rest. But do let me know if you have any other concerns, of course.

Thanks a alot!

kendy commented:
@ash: Well, even with the explanation, this sounds more like a workaround, rather than a root cause fix ;-)  I wonder - why do we end up in the state that loolwsd thinks the document is modified, but it actually is not in the first place?  Where have we lost the statuschanged message that was announcing the change from modified to non-modified?

I have this theory:

When you have a look where we trigger the success:false,result:unmodified, you'll see it is coming from desktop/source/lib/init.cxx .  We have code that I see as very problematic: Normally, all the save operation is async - we trigger ".uno:Save" and then have an interaction handler that answers with the state accordingly.

But here we synchronously shortcut this, without caring what might be in the event queue - and if it is !pDocSh->IsModified() [at **this** very moment], we just trigger the success:false,result:unmodified, regardless there might be other things in the event queue [ie. still in the need of processing, before we actually get to the .uno:Save].

@ash: Can you experiment with this & see how exactly that happens that the loolwsd and core state get out of sync?  Can you try to revert b91bb2f2702a82255969ae1da298159ee109d14a and implement the save skipping so that we get the "unmodified" state sent via the InteractionHandler, as part of the normal, proper .uno:Save operation?  [Or do we even need this sending of this state in the current world?]

Thank you in advance!

ash commented:
I believe you wanted to tag a different person.

kendy commented:
@Ashod: Meant to tag you of course.

@ash: Terribly sorry! :-)

Ashod commented:
@kendy: good points. As discussed offline, will follow your leads. Also, agreed to keep `_isModified` as 100% reflecting the last Core value and never changing it.

Thanks!

Ashod commented:
Closing as this has already been superseded.
